httpcache.go

```diff
@@ -1,402 +1,407 @@
 	"github.com/caddyserver/caddy/v2/caddyconfig/httpcaddyfile"
 	"github.com/caddyserver/caddy/v2/modules/caddyhttp"
-	"github.com/pquerna/cachecontrol/cacheobject"
+	"github.com/darkweak/souin/configurationtypes"
+	"github.com/darkweak/souin/plugins"
+	"github.com/darkweak/souin/rfc"
 	"go.uber.org/zap"
 )
 
-const userAgent = "Caddy"
+type key string
 
-// config, db and dmap are global to preserve the cache between reloads and to share the same cache between all routes
-var (
-	cfg  *Config
-	db   *olric.Olric
-	dmap *olric.DMap
-)
+const getterContextCtxKey key = "getter_context"
+const moduleName = "cache"
+const moduleID caddy.ModuleID = "http.handlers." + moduleName
 
 func init() {
-	caddy.RegisterModule(Cache{})
-	httpcaddyfile.RegisterGlobalOption("cache", parseCaddyfileGlobalOption)
-	httpcaddyfile.RegisterHandlerDirective("cache", parseCaddyfileHandlerDirective)
+	caddy.RegisterModule(SouinCaddyPlugin{})
+	httpcaddyfile.RegisterGlobalOption(moduleName, parseCaddyfileGlobalOption)
+	httpcaddyfile.RegisterHandlerDirective(moduleName, parseCaddyfileHandlerDirective)
 }
 
-type Config struct {
-	// Path to the Olric's configuration file. See https://github.com/buraksezer/olric#client-server-mode.
-	OlricConfig string `json:"olric_config,omitempty"`
-}
-
-// Cache implements a simple distributed cache.
-//
-// Caches only GET and HEAD requests. Honors the Cache-Control: no-cache header.
-//
-// Still TODO:
-//
-// - Properly set olric options
-// - Eviction policies and API
-// - Use single cache per-process
-type Cache struct {
-	Config
-	logger *zap.Logger
+// SouinCaddyPlugin declaration.
+type SouinCaddyPlugin struct {
+	plugins.SouinBasePlugin
+	Configuration *Configuration
+	logger        *zap.Logger
+	LogLevel      string `json:"log_level,omitempty"`
+	bufPool       *sync.Pool
+	Headers       []string                         `json:"headers,omitempty"`
+	Badger        configurationtypes.CacheProvider `json:"badger,omitempty"`
+	Olric         configurationtypes.CacheProvider `json:"olric,omitempty"`
+	TTL           configurationtypes.Duration      `json:"ttl,omitempty"`
 }
 
 // CaddyModule returns the Caddy module information.
-func (Cache) CaddyModule() caddy.ModuleInfo {
+func (SouinCaddyPlugin) CaddyModule() caddy.ModuleInfo {
 	return caddy.ModuleInfo{
-		ID:  "http.handlers.cache",
-		New: func() caddy.Module { return new(Cache) },
+		ID:  moduleID,
+		New: func() caddy.Module { return new(SouinCaddyPlugin) },
 	}
 }
 
-// Provision provisions c.
-func (c *Cache) Provision(ctx caddy.Context) error {
-	c.logger = ctx.Logger(c)
+type getterContext struct {
+	rw   http.ResponseWriter
+	req  *http.Request
+	next caddyhttp.Handler
+}
 
-	if db != nil {
-		// the cache is immutable, provision it only one time
-		return nil
+// ServeHTTP implements caddyhttp.MiddlewareHandler.
+func (s *SouinCaddyPlugin) ServeHTTP(rw http.ResponseWriter, req *http.Request, next caddyhttp.Handler) error {
+	if !plugins.CanHandle(req, s.Retriever) {
```

Comment:
More cases can be excluded in `plugin.CanHandle()` isn't it? Maybe could you use the `cachecontrol` library as we planed to in this module in this function as it's already a dependency of Souin? This will prevent the allocation potentially done by the buffer pool.

Reply:
You should also set the `Cache-Status` header here to hint that it's a miss.

Reply:
You're right, I'll update the core asap.

Reply:
So this won't prevent the allocation done by the buffer pool because we have to attach the cache-status on the response